### PR TITLE
CommonClient: Add a hints tab

### PR DESCRIPTION
**Pull request by Silvris**
## What is this fixing or adding?
Adds a tab to all CommonClient-inherited clients that displays any currently received hints.

## How was this tested?
Manually, by use during several Archipelago sessions.

No automated testing was added.

## If this makes graphical changes, please attach screenshots.
image

### CODE REVIEW

**Review by Zunawe — APPROVED** [review]
Messed around with it for a bit on BizHawkClient and everything seems fine. Nothing in the code really stands out to me as needing changing, but I didn't look too in depth and I'm not familiar with the UI code at all.

Where we're touching the line anyway, replacing single quote strings with double quote strings might be nice.

And a very minor suggestion for the column headers would be to remove the colons, since they already look appropriately like headers anyway.

**Comment by alwaysintreble** [comment]
> And a very minor suggestion for the column headers would be to remove the colons, since they already look appropriately like headers anyway.

could the headers maybe be slightly higher point to differentiate them a bit? or a line under them like a traditional table. i think it's fine as is, but that would look a bit nicer.

**Comment by Silvris** [comment]
>     * When hints are added the line height can become too small and the text overlaps.

This was an issue for resizing, definitely a bit mad that it's still showing up for some reason. Kivy will just forget to update when texture_size changes for some reason.

File: kvui.py [hunk 1]
```diff
@@ -121,6 +122,60 @@ class HovererableLabel(HoverBehavior, Label):
     pass
 
 
+class TooltipLabel(HovererableLabel):
+    tooltip = None
+
+    def create_tooltip(self, text, x, y):
+        text = text.replace("<br>", "\n").replace('&amp;', '&').replace('&bl;', '[').replace('&br;', ']')
```
[the remark this answers is not in the document]
Comment (ThePhar) [reply kvui.py:129]: Yes, if we can let's update any old code we touch to match style guide if possible.

File: kvui.py [hunk 2]
```diff
@@ -282,6 +286,64 @@ def apply_selection(self, rv, index, is_selected):
...
+        self.index = index
+        if "select" in data and not data['select'] and index not in self.no_select:
+            self.no_select.append(index)
+        self.striped = data['striped']
+        self.receiving_text = data['receiving']['text']
+        self.item_text = data['item']['text']
+        self.finding_text = data['finding']['text']
+        self.location_text = data['location']['text']
+        self.entrance_text = data['entrance']['text']
+        self.found_text = data['found']['text']
```
Comment (ThePhar) [inline kvui.py:319]: Please change `'` usages to `"`.

File: kvui.py [hunk 3]
```diff
@@ -282,6 +286,64 @@ def apply_selection(self, rv, index, is_selected):
...
+                if self.selected:
+                    self.parent.clear_selection()
+                else:
+                    text = "".join([self.receiving_text, "\'s ", self.item_text, " is at ", self.location_text, " in ",
+                                    self.finding_text, "\'s World", (" at " + self.entrance_text)
+                                    if self.entrance_text != "Vanilla"
+                                    else "", ". (", self.found_text.lower(), ")"])
+                    temp = MarkupLabel(text).markup
+                    text = "".join(part for part in temp if not part.startswith(("[color", "[/color]", "[ref=", "[/ref]")))
+                    Clipboard.copy(escape_markup(text).replace('&amp;', '&').replace('&bl;', '[').replace('&br;', ']'))
```
Comment (ThePhar) [inline kvui.py:338]: Ditto.

File: kvui.py [hunk 4]
```diff
@@ -552,6 +623,40 @@ def fix_heights(self):
...
+                'receiving': {'text': self.parser.handle_node({'type': "player_id", 'text': hint['receiving_player']})},
+                'item': {'text': self.parser.handle_node({'type': "item_id", 'text': hint['item'], 'flags': hint['item_flags']})},
+                'finding': {'text': self.parser.handle_node({'type': "player_id", 'text': hint['finding_player']})},
+                'location': {'text': self.parser.handle_node({'type': "location_id", 'text': hint['location']})},
+                'entrance': {'text': self.parser.handle_node({'type': "color" if hint['entrance'] else "text",
+                                                         'color': "blue", 'text': hint['entrance']
+                                                         if hint['entrance'] else "Vanilla"})},
+                'found': {'text': self.parser.handle_node({'type': "color", 'color': "green" if hint['found'] else "red",
+                                                      'text': "Found" if hint['found'] else "Not Found"})},
+            })
```
Comment (ThePhar) [inline kvui.py:657]: Ditto.

File: kvui.py [hunk 5]
```diff
@@ -627,4 +732,3 @@ def _handle_text(self, node: JSONMessagePart):
 if os.path.exists(user_file):
     logging.info("Loading user.kv into builder.")
     Builder.load_file(user_file)
-
```
Comment (ThePhar) [inline kvui.py:630]: Python files should end with a newline character.

**Review by ThePhar — CHANGES_REQUESTED** [review]
Looks fine to me if we can update the quote usage. 

Has this been tested in `CommonClient`-derived clients that also add additional tabs like Starcraft II client to make sure it renders okay?

File: kvui.py (hunk 5 above)
Comment (Silvris) [reply kvui.py:630]: This was an extra newline; the file still ends with a newline as per the style guide.

**Comment by Silvris** [comment]
image
Updated visuals based on feedback from Discord

**Comment by Silvris** [comment]
> Has this been tested in `CommonClient`-derived clients that also add additional tabs like Starcraft II client to make sure it renders okay?

I have not completely tested SC2 client (I do not have SC2 downloaded), but I did load it to ensure the hint tab was correctly placed and did not interfere with their additions. I'll need to confirm LADX,  but I cannot confirm Wargroove as I do not own it.

**Review by ThePhar — APPROVED** [review]

**Comment by ThePhar** [comment]
Just checked LADX and Wargroove, they looked fine on my machine.

**Pull request merged**